src/ump.rs

```diff
@@ -537,6 +693,82 @@ fn channel_voice1(bytes: [u8; 4]) -> MidiMessage {
     }
 }
 
+fn flex(bytes: [u8; 16]) -> MidiMessage {
+    let group = bytes[0] & 0x0f;
+    let channel_or_group = bytes[1] & 0x0f;
+    let address = bytes[1] >> 4;
+    let address: FlexAddress = match address {
+        0 => FlexAddress::Channel(channel_or_group),
+        1 => FlexAddress::Group(channel_or_group),
+        2 => FlexAddress::_Reserved1(channel_or_group),
+        3 => FlexAddress::_Reserved2(channel_or_group),
+        _ => unreachable!(),
+    };
+    let status = bytes[3];
+    match status {
+        0x0 => {
+            use core::convert::TryInto;
+            let (ns_per_qn_bytes, rest) = bytes.split_at(core::mem::size_of::<u32>());
+            let ns_per_qn = u32::from_le_bytes(ns_per_qn_bytes.try_into().unwrap());
```

Comment:
Does the spec require this to be little-endian?

Reply:
That's a good question. It doesn't actually specify the endianness, leaving it up to the implementation.

From M2-104-UM UMP Format & MIDI 2.0 Protocol v1.1.1 2023-07-19, p. 20-21:
> the formats of such byte streams, including the byte order to be used for such transport and storage, are outside the scope of this specification. [...] separate transport specifications will define formats and byte orders for each particular transport, and separate file format specifications addressing the UMP Format will define byte orders for each particular file format. For the internals of any given implementation, a device or system may use any desired format, including native-
endian 32-bit words.
> MMA/AMEI expect that for every standardized transport that uses the UMP Format, a separate specification will exist to define how to carry UMP payload data for that standardized transport.

The specification itself draws diagrams of UMP messages in little-endian.
Only MIDI Clip File specification strongly defines its endianness (which is big-endian).

According to [the Linux kernel docs](https://docs.kernel.org/sound/designs/midi-2.0.html):
> The 32-bit words in the UMP packet payload are always in CPU native endianness. Transport drivers are responsible to convert UMP words from / to system endianness to required transport endianness / byte order.

So I kind of just assumed desktop computers. But indeed, considering that this lib is no-std which, I guess, means it also presumes embedded systems as its target, my assumption was not entirely correct.

Reply:
That's what I expected, in which case we should use `from_ne_bytes` here instead of assuming little endianness. 

I have a pending refactor that changes the underlying representation of everything to use native-endian 4 byte words for all UMPs and messages for compatibility with MacOS, Windows, and Linux(ALSA) MIDI 2 APIs (this crate was written before all those were published).

Reply:
Good, thanks for the comment, makes perfect sense. I didn't know there was a native-endian option!
I've been thinking about doing this too, so I think you'll be seeing me join in to help you out.

Reply:
I will push my branch later today!